fetch.bs

```diff
@@ -2520,6 +2726,100 @@ consideration for the security consequences. New exceptions can be proposed by
 <a href=https://github.com/whatwg/fetch/issues/new>filing an issue</a>.
 
 
+<h3 id=content-type-header>`<code>Content-Type</code>` header</h3>
+
+<p>The `<code>Content-Type</code>` header is largely defined in HTTP. Its processing model is
+defined here as the ABNF defined in HTTP is not compatible with web content. [[HTTP]]
+
+<p>To
+<dfn export for="header list" lt="extract a MIME type|extracting a MIME type" id=concept-header-extract-mime-type>extract a MIME type</dfn>
+from a <a for=/>header list</a> <var>headers</var>, run these steps:
+
+<ol>
+ <li><p>Let <var>charset</var> be null.
+
+ <li><p>Let <var>essence</var> be null.
+
+ <li><p>Let <var>mimeType</var> be null.
+
+ <li><p>Let <var>values</var> be the result of
+ <a for="header list">getting, decoding, and splitting</a> `<code>Content-Type</code>` from
+ <var>headers</var>.
+
+ <li><p>If <var>values</var> is null, then return failure.
+
+ <li>
+  <p><a for=list>For each</a> <var>value</var> of <var>values</var>:
+
+  <ol>
+   <li><p>Set <var>mimeType</var> to the result of <a lt="parse a MIME type">parsing</a>
+   <var>value</var>.
+
+   <li><p>If <var>mimeType</var> is failure, then <a for=iteration>continue</a>.
+
+   <li>
+    <p>If <var>mimeType</var>'s <a for="MIME type">essence</a> is not "<code>*/*</code>" or
+    <var>essence</var>, then:
+
+    <ol>
+     <li><p>Set <var>charset</var> to null.
+
+     <li><p>If <var>mimeType</var>'s <a for="MIME type">parameters</a>["<code>charset</code>"]
```

Comment:
One other question:  Should we be taking the first MIME type rather than the last?  This would change behavior, but it would match the fact we prefer earlier charsets.  I had thought this was discussed as a potential change elsewhere.  I'm not pushing strongly for a change here, as I think all browsers currently prefer later MIME types, but thought I'd bring it up.

Reply:
So yeah, if you think that's doable it's certainly worth considering. But what if it's `*/*` or fails to parse, would we then still go to the next? If so, I'm not sure what it would buy us in terms of security.

Reply:
Not sure that skipping over invalid MIME types really costs us much in terms of security, but I'd be perfectly fine with just taking the first line, and ignoring the rest.  Downside of that is, as with most every decision here, we just don't know how much of the web that breaks.

Reply:
So to be clear, we'd combine, split, and then parse? As opposed to combine and then parse (which we do on the request side for CORS)?

Reply:
We're already combining, splitting, and then parsing, actually.  The "For each value of values" would just be removed, and we'd use value of [0] everywhere.  So it's one less step in the code, which I'm not sure can really be described as splitting, if what we were doing before isn't splitting.

Sorry, I'm not all that familiar with CORS.  Which section are you referring to as the request side of CORS that works differently?

Reply:
To be specific, the `Content-Type` parser defined at https://whatpr.org/fetch/831.html#cors-safelisted-request-header for use on requests fails on `text/plain,application/json` as it doesn't split. (And this is intentional for security reasons.)

Anyway, it would indeed be a rather big simplification if we'd simply use the first value as that would also remove the need for `charset` moving and such. I'm rather skeptical this'll work out, but if you want to try it I guess I'm not opposed.

Reply:
The more I think about it, the more I think you're probably right that just taking the first would break too much.

Seems to me like the safest way to CORS-safelist-check check the Content-Type request header would be to either check all MIME types, not just the first one, or refuse to send any but the first.  That way, we wouldn't have to rely on servers implementing a fully compliant Content-Type parser.  But that's perhaps getting a bit far afield.

Reply:
https://github.com/whatwg/fetch/issues/838